### ISSUE TITLE
Object table null - please_report_this_error_to_help_at_realm_dot_io

### ISSUE THREAD

**Issue by plivesey** [realm/realm-cocoa#4824]
## Goals
I want Realm to not crash in this scenario.

## Expected Results
I expect realm to not crash.

## Actual Results

Stack trace:
```
table.cpp:4423: [realm-core-2.3.3] Assertion failed: &src_table->get_column_link_base(src_col_ndx).get_target_table() == this
0   Realm                               0x0000000101e9dcec _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 44
1   Realm                               0x0000000101e9df3c _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 388
2   Realm                               0x0000000101fe4c00 _ZN5realm5Table17get_backlink_viewEmPS0_m + 180
3   Realm                               0x0000000101c4b6ec _ZL20RLMGetLinkingObjectsP13RLMObjectBaseP11RLMProperty + 376
4   Realm                               0x0000000101c4ac18 RLMDynamicGet + 404
5   RealmSwift                          0x0000000102463cbc _TFC10RealmSwift18LinkingObjectsBaseg10rlmResultsGCSo10RLMResultsCSo9RLMObject_ + 356
6   RealmSwift                          0x00000001024654c8 _TFC10RealmSwift14LinkingObjectsg5firstGSqx_ + 76
7   Superhuman                          0x00000001003a54c8 _TFC10Superhuman12MessageModelg6threadGSqCS_11ThreadModel_ + 64
```

Function which crashes:
```
static inline RLMLinkingObjects *RLMGetLinkingObjects(__unsafe_unretained RLMObjectBase *const obj, __unsafe_unretained RLMProperty *const property) {
    auto& objectInfo = obj->_realm->_info[property.objectClassName];
    auto linkingProperty = objectInfo.objectSchema->property_for_name(property.linkOriginPropertyName.UTF8String);
    auto backlinkView = obj->_row.get_table()->get_backlink_view(obj->_row.get_index(), objectInfo.table(), linkingProperty->table_column);
    realm::Results results(obj->_realm->_realm, std::move(backlinkView));
    return [RLMLinkingObjects resultsWithObjectInfo:objectInfo results:std::move(results)];
}
```

and the line:
`auto backlinkView = obj->_row.get_table()->get_backlink_view(obj->_row.get_index(), objectInfo.table(), linkingProperty->table_column);`

And here is some debug information printed right before the crash occurs:
```
(lldb) print obj
(RLM:Managed 6 MessageModel *) $3 = 0x0000000116831c50
(lldb) print obj->_row
(realm::Row) $4 = {
  realm::RowBase = {
    m_table = {
      realm::util::bind_ptr<realm::Table> = {
        m_ptr = 0x0000000000000000
      }
    }
    m_row_ndx = 356
    m_prev = 0x0000000000000000
    m_next = 0x0000000000000000
  }
}
(lldb) print obj->_row.get_table()
(table_type *) $5 = 0x0000000000000000
(lldb) print obj->_row.get_index()
(size_t) $6 = 356
(lldb) print objectInfo.table()
(realm::Table *) $7 = 0x00000001038be400
(lldb) print linkingProperty->table_column
(size_t) $8 = 9
```

## Steps to Reproduce
We can 100% reproduce this problem on our app with a certain user. However, only this user seems to cause the crash. If it would help, I can send you the project, but can't send you the user's credentials. I could come by the office and reproduce in house if it'd help you.

Let me know if the project would help you, I can email it over.

## Code Sample
```
            // above: let discardedMessageIds: [String]
            let realm = try! Realm(configuration: Realm.configurationForUser(userEmail))
            let messages = discardedMessageIds.flatMap { realm.object(ofType: MessageModel.self, forPrimaryKey: $0) }

            messages.forEach { message in
                guard let thread = message.thread else { return } // This line crashes
```

```
class MessageModel: Object {
    override static func primaryKey() -> String? {
        return "id"
    }
    
    override static func ignoredProperties() -> [String] {
        return ["contentZoom", "labelIds", "references", "superhumanDraftId"]
    }
    
    override static func indexedProperties() -> [String] {
        return ["dateSent"]
    }

    dynamic var id = ""
    dynamic var ... // a bunch of properties

    var thread: ThreadModel? { return threads.first }
    fileprivate let threads = LinkingObjects(fromType: ThreadModel.self, property: "messages")

    //...
}
```

## Work around

The message appears to be invalidated. If we guard against that it 'fixes' the bug. However, it's unclear why this was invalidated (we're not sure it should have been) and also, given the code above, there is no point where this could have been invalidated.

## Version of Realm and Tooling
```
ProductName:	Mac OS X
ProductVersion:	10.12.4
BuildVersion:	16E195

/Applications/Xcode.app/Contents/Developer
Xcode 8.3
Build version 8E162

/usr/local/bin/pod
1.2.0
Realm (2.4.4)
RealmSwift (2.4.4)
RealmSwift (~> 2.4)

/bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin16)

/usr/local/bin/carthage
0.18.1
(not in use here)

/usr/bin/git
git version 2.11.0 (Apple Git-81)
```

**Comment by plivesey** [comment]
I debugged this with @jpsim and found two issues:
1 - we were writing to realm in a loop:
```
let objects: [RealmObject]
for object in objects {
    realm.write {
        // do something
    }
}
```
Not only is this inefficient, but it also means we do a realm refresh. This means that while the loop iterates, the objects could be deleted.

2 - We had duplicate objects in this array. This meant that when we deleted things, we sometimes did double deletes.

**Comment by bdash** [comment]
Seems like this should be failing in a more understandable way than an assertion failure down in core? Worth keeping open to improve that?

**Comment by plivesey** [comment]
@jpsim suggested that this would provide a better assertion in 2.5.1 (or maybe just on master). So I think this should already be fixed? But feel free to reopen if you think that's worthwhile until it's verified.

**Comment by jpsim** [comment]
Yes, the exception with more understandable message was merged in #4817
